### ISSUE TITLE
build-arg is empty in RUN when defined before FROM

### ISSUE THREAD

**Issue by fugkco** [containers/buildah#2855]
**Description**
I'm unable to use simple build args within my Dockerfile except for usage in the `FROM` directive.

**Steps to reproduce the issue:**
```
$ cat Dockerfile
ARG MY_APP_VERSION="123"
FROM alpine:3.12
RUN set -eux; echo $MY_APP_VERSION
```

**Describe the results you received:**
```
$ podman build -t test .
STEP 1: FROM alpine:3.12
STEP 2: RUN set -eux; echo $MY_APP_VERSION
/bin/sh: MY_APP_VERSION: parameter not set
Error: error building at STEP "RUN set -eux; echo $MY_APP_VERSION": error while running runtime: exit status 2
```

**Describe the results you expected:**
The ARG `MY_APP_VERSION` should be detected

This works fine when ARG is _after_ `FROM` directive:
```
$ cat Dockerfile
FROM alpine:3.12
ARG MY_APP_VERSION="123"
RUN set -eux; echo $MY_APP_VERSION > /wot

$ podman build -t test .
STEP 1: FROM alpine:3.12
STEP 2: ARG MY_APP_VERSION="123"
--> Using cache 88475f7de7c94ae206d2c8aa960a4107a6ab92f40d5b45786cdf76fdbfd1f528
--> 88475f7de7c
STEP 3: RUN set -eux; echo $MY_APP_VERSION > /wot
+ echo 123
STEP 4: COMMIT test
--> faecb03aea7
faecb03aea70dbd222721df9a8dcd4627d240966be0994db804feea3a3b2ca75
```

**Output of `podman version` if reporting a `podman build` issue:**

```
$ podman version
Version:      2.1.1
API Version:  2.0.0
Go Version:   go1.15.2
Built:        Thu Jan  1 01:00:00 1970
OS/Arch:      linux/amd64

```

**Output of `cat /etc/*release`:**

```
$ cat /etc/os-release
NAME="Pop!_OS"
VERSION="20.10"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 20.10"
VERSION_ID="20.10"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=groovy
UBUNTU_CODENAME=groovy
LOGO=distributor-logo-pop-os

```

**Output of `uname -a`:**

```
$ uname -a
Linux pop-os 5.8.0-7630-generic #32~1605108806~20.10~7e52b13-Ubuntu SMP Wed Nov 11 19:10:30 UTC  x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
$ cat /etc/containers/storage.conf | grep -v '^#' | grep .
[storage]
driver = ""
runroot = "/var/run/containers/storage"
graphroot = "/var/lib/containers/storage"
[storage.options]
additionalimagestores = [
]
[storage.options.overlay]
mountopt = "nodev"
[storage.options.thinpool]
```

**Comment by TomSweeneyRedHat** [comment]
@fugkco Thanks for the issue report.  I'll look tomorrow when the eyelids are less heavy, but I think this may be working as anticipated.  I seem to remember that the ENV's before the first FROM are discarded at that point for some reason, but need to dive further.

**Comment by TomSweeneyRedHat** [comment]
@fugkco I've verified in the code and also in this nice blog [post](https://benkyriakou.com/posts/docker-args-empty).  This is working as expected.  The one that you gave as an example that works:

```
$ cat Dockerfile
FROM alpine:3.12
ARG MY_APP_VERSION="123"
RUN set -eux; echo $MY_APP_VERSION > /wot
```
This is the way that the ARG should be defined in the Dockerfile.  That is if you want to use an ARG in your Dockerfile, it needs to be defined after the FROM in which it will be used and before any following FROM statements if there are multiple FROM statements.

Closing the issue.
